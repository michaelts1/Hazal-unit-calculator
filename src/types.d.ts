export interface Unit {
	hidden: boolean
	name: string
	value: number
}

export type Ruler = 'גרח' | 'חזון_איש'

export type Page = 'length' | 'volume'

export type DefaultUnits = {
	input: string,
	output: string,
}
