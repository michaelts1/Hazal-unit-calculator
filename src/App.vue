<script lang="ts">
import { defineComponent } from 'vue'
import LengthPage from './components/LengthPage.vue'

export default defineComponent({
	components: {
		LengthPage,
	},

	data() {
		return {
			view: 'LengthPage',
		}
	},
})
</script>

<template>
	<h1>מחשבון יחידות תלמודיות</h1>
	<LengthPage v-if="view === 'LengthPage'" />

	<footer>© כל הזכויות שמורות למיכאל צבאן</footer>
</template>
