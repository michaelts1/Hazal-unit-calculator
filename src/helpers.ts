/** Rounds to 10 digits */
export const roundNum = (num: number) => Math.round(num * 1e10) / 1e10
