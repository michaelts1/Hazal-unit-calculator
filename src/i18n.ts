import { reactive } from 'vue'

const messages = {
	en: {
		_dir: 'ltr',
		_name: 'English',
		amah: 'amah',
		beitza: 'beitza',
		centimeterLong: 'centimeter',
		centimeterShort: 'cm',
		chazonIsh: 'Chazon Ish',
		copyright: '© Michael Tsaban',
		cubicCentimeterShort: 'cc',
		cubicMeterShort: 'cbm',
		eifa: 'eifa',
		etzba: 'etzba',
		grach: 'Rav Chaim Naeh',
		grogeret: 'grogeret',
		input: 'Input',
		invalidNumber: 'Please enter a number',
		kav: 'kav',
		kilometerLong: 'kilometer',
		kilometerShort: 'km',
		kor: 'kor',
		kortov: 'kortov',
		length: 'Length',
		literLong: 'liter',
		litra: 'litra',
		log: 'log',
		messageLength: 'Attention: The sit is calculated according to Maimonides',
		messageVolume: 'Attention: The peras is calculated according to Maimonide and the zait according to the Shulchan Aruch',
		meterLong: 'meter',
		mil: 'mil',
		millimeterShort: 'mm',
		output: 'Output',
		parsa: 'parsa',
		peras: 'peras',
		posek: 'Posek',
		ris: 'ris',
		seah: 'seah',
		sit: 'sit',
		switchInputOutput: 'Switch input and output',
		tefach: 'tefach',
		title: 'Hazal Unit Calculator',
		type: 'Type',
		volume: 'Volume',
		zait: 'zait',
		zeret: 'zeret',
	},

	he: {
		_dir: 'rtl',
		_name: 'עברית',
		amah: 'אמה',
		beitza: 'ביצה',
		centimeterLong: 'סנטימטר',
		centimeterShort: 'ס"מ',
		chazonIsh: 'חזון איש',
		copyright: '© מיכאל צבאן',
		cubicCentimeterShort: 'סמ"ק',
		cubicMeterShort: 'מ"ק',
		eifa: 'איפה',
		etzba: 'אצבע',
		grach: 'רב חיים נאה',
		grogeret: 'גרוגרת',
		input: 'קלט',
		invalidNumber: 'נא הכנס מספר',
		kav: 'קב',
		kilometerLong: 'קילומטר',
		kilometerShort: 'ק"מ',
		kor: 'כור',
		kortov: 'קורטוב',
		length: 'מרחק',
		literLong: 'ליטר',
		litra: 'ליטרא',
		log: 'לוג',
		messageLength: 'שים לב: הסיט מחושב לפי שיטת הרמב"ם',
		messageVolume: 'שים לב: הפרס מחושב לפי שיטת הרמב"ם והזית לפי השולחן ערוך',
		meterLong: 'מטר',
		mil: 'מיל',
		millimeterShort: 'מ"מ',
		output: 'תוצאה',
		parsa: 'פרסה',
		peras: 'פרס',
		posek: 'שיטת פסיקה',
		ris: 'ריס',
		seah: 'סאה',
		sit: 'סיט',
		switchInputOutput: 'החלף בין הקלט לתוצאה',
		tefach: 'טפח',
		title: 'מחשבון יחידות תלמודיות',
		type: 'תחום',
		volume: 'נפח',
		zait: 'זית',
		zeret: 'זרת',
	},
}

type MessagesLocale = keyof typeof messages
type Messages = typeof messages[MessagesLocale]
type MessageName = keyof Messages

class I18n {
	currentLocale: MessagesLocale
	messages!: Messages

	constructor(locale: MessagesLocale) {
		this.currentLocale = locale
		this.changeLocale(locale)
	}

	t(name: MessageName, locale?: MessagesLocale) {
		return locale ? messages[locale][name] : this.messages[name]
	}

	changeLocale(locale: MessagesLocale) {
		this.currentLocale = locale
		localStorage.setItem('locale', locale)
		this.messages = messages[this.currentLocale]

		document.documentElement.dir = this.t('_dir')
		document.title = this.t('title')
	}
}

/** @returns {MessagesLocale} Last used locale, defaults to 'he' */
function getLocaleCookie(): MessagesLocale {
	const locale = localStorage.getItem('locale') as MessagesLocale | null
	return locale ?? 'he'
}

export const i18n = reactive(new I18n(getLocaleCookie()))
