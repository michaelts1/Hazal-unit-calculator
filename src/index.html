<!doctype html>
<html>
	<head>
		<title>מחשבון יחידות תלמודיות</title>
		<link rel="icon" href="icon.png">
		<link rel="stylesheet" href="css/index.css">
	</head>
	<body>
		<div id="app"></div>

		<script src="index.js"></script>
	</body>
</html>
