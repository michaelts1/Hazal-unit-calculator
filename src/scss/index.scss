@mixin table-border {
	border-collapse: collapse;
	border: 1px solid black;
}

@mixin title($color) {
	text-align: center;
	color: $color;
}

body {
	background: #f9f9f9;
	direction: rtl;
	font-family: Arial, Helvetica, sans-serif;
	margin: 0;
}

header {
	border-bottom: 2px solid #ddd;
	margin: 0 0 2em;
	padding: 0;
	width: 100%;
}

footer {
	bottom: 5px;
	font-size: small;
	left: 0;
	position: fixed;
	right: 0;
	text-align: center;
}

h1 {
	@include title(green);
}

h6 {
	@include title(#777);
	margin: 2em 0 1em 0;
}


table {
	@include table-border;
}

th, td {
	@include table-border;
	padding: 0.25em;
	text-align: center;
	width: 75px;
}

article {
	margin: 0 auto;
	width: fit-content;
}
