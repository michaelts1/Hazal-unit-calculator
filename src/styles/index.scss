$min-table-length: 1085px;
$low-width-border: 1px solid #999;

@mixin title($color) {
	text-align: center;
	color: $color;
}

@mixin background {
	background: #f9f9f9;
}

@mixin cell-background($parity) {
	@if $parity == odd {
		background: #fbfbfb;
	} @else if $parity == even {
		background: #f0f0f0;
	}
}

html[dir='ltr'] {
	#table-low-width .col:nth-of-type(1) {
		border-right: none !important;
		border-left: $low-width-border;
	}

	select {
		padding-left: 8px !important;
	}

	#switch-lang-container {
		@media (max-width: 380px) {
			direction: rtl !important;
		}
	}
}

body {
	@include background;

	font-family: Arial, Helvetica, sans-serif;
	margin: 0;
	min-width: 337px;
}

#bottom {
	@include background;

	bottom: 0;
	height: 25px;
	left: 0;
	position: fixed;
	right: 0;

	footer {
		bottom: 5px;
		font-size: small;
		text-align: center;
		width: 100%;
	}
}

#app { // Nest all styles inside an id to override boostrap - https://stackoverflow.com/a/13029669
	max-width: unset !important;
	margin-bottom: 25px;

	header {
		border-bottom: 2px solid #ddd;
		position: relative;
		margin: 0 0 2em;
		padding: 0;
		width: 100%;
	}

	h1 {
		@include title(green);
		margin: 0;
		padding: .25rem 0;
	}

	h5 {
		@include title(darkgreen);
		margin: 0;
	}

	h6 {
		@include title(#666);
		font-size: 8pt;
		font-weight: bold;
		margin: 1.5em 5em 1em;
	}

	select {
		background-color: #f5fbf5;
		padding: 4px 30px 4px 0;
	}

	th, td {
		@media (max-width: 1200px) {
			padding: .25em;
			width: 96px;
		}

		border-collapse: collapse;
		border: 1px solid #888;
		padding: .25em .5em;
		text-align: center;
		text-transform: capitalize;
		width: 104px;

		&:nth-of-type(odd) {
			@include cell-background(odd);
		}

		&:nth-of-type(even) {
			@include cell-background(even);
		}
	}

	.btn-light {
		border-color: #c8c8c8;
		padding: 0;

		img {
			margin: 6px;
			width: 15px;
		}
	}

	.container {
		max-width: unset !important;
		margin-bottom: 25px;

		@media (max-width: 500px) {
			padding: 0;
		}
	}

	.input-amount {
		direction: ltr;
		margin: .5em 0 1em;
		width: 130px;

		&.invalid {
			box-shadow: #b88 1px 1px 3px 2px;

			&:focus {
				box-shadow: #b88 0 -1px 3px 5px;
			}
		}

		&:disabled {
			background: #f5f5f5;
			color: black;
		}
	}

	.mx-auto {
		width: fit-content;

		.select {
			@media (max-width: 768px) {
				margin: 0 1.5em;
			}

			@media (max-width: 576px) {
				margin: 0 .5em;
			}

			@media (max-width: 400px) {
				display: block;
				margin: .5em;
			}

			margin: 0 2em;
		}
	}

	.wrapper-row {
		display: flex;
		margin: 1em auto;
		width: fit-content;
	}

	.wrapper-column {
		align-items: center;
		display: flex;
		flex-direction: column;
		justify-content: center;
		padding: 0 1em;

		@media (min-width: $min-table-length) {
			.input-amount, .form-select {
				display: inline-block;
				width: 130px;
				margin: 10px;
			}
		}

		@media (max-width: 500px) {
			padding: 0 .5em;
		}
	}

	#page-select, #ruler-select {
		display: inline-block;
		margin: 0 5px;
		width: unset;
	}

	#switch-lang-container {
		bottom: 0;
		left: 10px;
		padding: 5px 10px;
		position: absolute;
		top: 0;

		@media (max-width: 440px) {
			padding: 5px;
		}

		@media (max-width: 415px) {
			padding: 5px 0;
		}

		@media (max-width: 400px) {
			left: 5px;
		}

		@media (max-width: 380px) {
			direction: ltr;
		}
	}

	#switch-lang {
		font-size: 11pt;
		padding: 0;
		position: relative;
		text-transform: capitalize;
		top: 50%;
		transform: translateY(-50%);

		@media (max-width: 380px) {
			width: 3ch;
			overflow: hidden;
			text-overflow: '';
		}
	}

	#table-low-width {
		width: 250px;

		.row {
			&:nth-of-type(odd) {
				@include cell-background(odd);
			}

			&:nth-of-type(even) {
				@include cell-background(even);
			}

			&:nth-last-of-type(1) {
				border-bottom: $low-width-border;
			}
		}

		.col {
			border-top: $low-width-border;
			border-right: $low-width-border;
			display: table-cell;
			padding: 5px 10px;
			text-align: center;

			&:nth-of-type(1) {
				font-weight: bold;
				text-transform: capitalize;
			}

			&:nth-of-type(2) {
				border-left: $low-width-border;
			}
		}
	}
}
