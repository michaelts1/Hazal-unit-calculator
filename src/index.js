import * as Vue from 'vue'
// @ts-ignore not sure why, but I am getting 'Cannot find module' error. It still works though.
import App from './App.vue'

Vue.createApp(App).mount('#app')
