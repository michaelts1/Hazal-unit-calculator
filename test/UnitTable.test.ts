import { describe, expect, it } from 'vitest'
import { mount } from '@vue/test-utils'
import { roundNum } from '../src/helpers'
import type { Unit } from '../src/types'
import UnitTable from '../src/components/UnitTable.vue'

describe('Component UnitTable', () => {
	it('exists', () => {
		expect(UnitTable).toBeTruthy()
	})

	const units: Unit[] = [
		{ hidden: false, name: 'אצבע', value: .005 },
		{ hidden: false, name: 'סנטימטר', value: .1 },
		{ hidden: false, name: 'טפח', value: 4 },
		{ hidden: false, name: 'זרת', value: 4000 },
	]
	const wrapper = mount(UnitTable, {
		props: {
			units,
			format: function formatNum(num: number): string {
				let unit = ' מטר'

				if (num < .01) {
					num *= 1000
					unit = ' מ"מ'
				} else if (num < 1) {
					num *= 100
					unit = ' ס"מ'
				} else if (num > 1000) {
					num /= 1000
					unit = ' ק"מ'
				}

				return roundNum(num) + unit
			},
		},
	})

	it('has a table with the correct number of columns', () => {
		expect(wrapper.find('table').exists()).toBe(true)
		expect(wrapper.findAll('th')).toHaveLength(units.length)
	})

	it('matches snapshot', () => {
		expect(wrapper.html()).toMatchSnapshot()
	})

	it('lists the correct values and reacts to value changes', async () => {
		expect(wrapper.text()).toContain('5 מ"מ')
		expect(wrapper.text()).toContain('4 מטר')
		expect(wrapper.text()).toContain('4 ק"מ')

		units[0].value = .5
		await wrapper.vm.$forceUpdate() // $forceUpdate is async (the type definition is wrong)
		expect(wrapper.text()).not.toContain('5 מ"מ')
		expect(wrapper.text()).toContain('50 ס"מ')
	})
})
