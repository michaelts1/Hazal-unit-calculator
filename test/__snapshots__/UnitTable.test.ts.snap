// Vitest Snapshot v1

exports[`Component UnitTable > matches snapshot 1`] = `
"<table data-v-2f1b50ac=\\"\\">
  <thead data-v-2f1b50ac=\\"\\">
    <tr data-v-2f1b50ac=\\"\\">
      <th data-v-2f1b50ac=\\"\\">אצבע</th>
      <th data-v-2f1b50ac=\\"\\">סנטימטר</th>
      <th data-v-2f1b50ac=\\"\\">טפח</th>
      <th data-v-2f1b50ac=\\"\\">זרת</th>
    </tr>
  </thead>
  <tbody data-v-2f1b50ac=\\"\\">
    <tr data-v-2f1b50ac=\\"\\">
      <td data-v-2f1b50ac=\\"\\">5 מ\\"מ</td>
      <td data-v-2f1b50ac=\\"\\">10 ס\\"מ</td>
      <td data-v-2f1b50ac=\\"\\">4 מטר</td>
      <td data-v-2f1b50ac=\\"\\">4 ק\\"מ</td>
    </tr>
  </tbody>
</table>"
`;
